### DISCUSSION TITLE
Undefined index when resolving state of a relationship

### DISCUSSION THREAD

**Discussion by aluisiora** [spatie/laravel-model-states#181] [Bugs]
Hi! After some time of troubleshooting, I couldn't solve this issue where, when resolving a "belongsTo" relationship where the current model has state but also the parent has states too, for some reason, the "StateCaster" tries to lookup for the parent states within the current model state property. Example:

I got an envelope that has many recipients, the bug occurs when trying to fetch the recipient with the envelope:
```php

namespace App\Models;

use App\Support\Recipient\States\EnvelopeState;
use Illuminate\Database\Eloquent\Model;
use Spatie\ModelStates\HasStates;

class Envelope extends Model
{
    use HasStates;

    protected $casts = [
        'state' => EnvelopeState::class
    ];

    public function recipients()
    {
        return $this->hasMany(Recipient::class);
    }
}
```

```php

namespace App\Models;

use App\Support\Recipient\States\RecipientState;
use Illuminate\Database\Eloquent\Model;
use Spatie\ModelStates\HasStates;

class Recipient extends Model
{
    use HasStates;

    protected $casts = [
        'state' => RecipientState::class
    ];

    public function envelope()
    {
        return $this->belongsTo(Envelope::class);
    }
}
```

Each model state is in a different directory:
```
app
|   Support
|---- Envelope
|------ States
|-------- Draft.php
|-------- EnvelopeState.php
|---- Recipient
|------ States
|-------- Queued.php
|-------- RecipientState.php
```

When I try to load and use both like this for example:
```php
$recipient = Recipient::with('envelope')
                        ->whereHas('envelope', fn($query) $query->where('id', $envelopeId))
                        ->findOrFail($recipientId);

if (! $recipient->envelope->state->equals(Draft::class)) {
     // .. do something
}

if (! $recipient->state->equals(Queued::class)) {
     // .. do something
}
```

It throws an error like this:
![Captura de tela 2021-07-26 123525](https://user-images.githubusercontent.com/14558910/127017213-ddf1acbd-637b-49b9-bd99-83ecf9eb2c97.png)

**Comment by aluisiora** [comment]
If anyone has this issue, my workaround was to load one model then the other:

```php
$envelope = Envelope::findOrFail($envelopeId);

if (! $envelope->state->equals(Draft::class)) {
     // .. do something
}

$recipient = Recipient:findOrFail($recipientId);

if (! $recipient->state->equals(Queued::class)) {
     // .. do something
}
```

**Comment by brendt** [comment]
Which version of the package is this?

**Reply by aluisiora** [reply]
This was on version 2.1.0

**Reply by brendt** [reply]
Feel free to submit a PR with a failing test so that I can look into it further.

**Comment by aluisiora** [comment]
I've managed to pin down the issue after looking into it further, it turned out be on my end. I thought it was related to a whereHas but it was actually because of a join.

I left some of the tests in my project breaking so I could come back to it, the real problem happens when joining tables, when I don't specify from which table I want the columns.

I use hybrid uuid, so in my code I have something like this:

```php
$recipient = Recipient::with('envelope')
                        ->join('envelopes', 'envelopes.id', '=', 'recipients.envelope_id')
                        ->where('envelopes.uuid', $envelopeUuid)
                        ->where('recipients.uuid', $recipientUuid)
                        ->firstOrFail(): // <-- the issue
```

Both of these tables had the same column `state` but had different casts. If anyone end up here with this same problem, when joining tables like this, you must specify from with table you want the columns from, otherwise the model won't know which `state` column is which since there are two (one from recipients table and another from envelopes table in this case).

The fix:
```php
$recipient = Recipient::with('envelope')
                        ->join('envelopes', 'envelopes.id', '=', 'recipients.envelope_id')
                        ->where('envelopes.uuid', $envelopeUuid)
                        ->where('recipients.uuid', $recipientUuid)
                        ->firstOrFail('recipients.*'): // <-- select the columns
```
